proposals/2020-07-01_aws-kms-keyring-redesign/background.md

```diff
@@ -0,0 +1,1396 @@
+3. Ensure the ESDK _attempts_ to decrypt my data in a parallel manner,
+   _attempting_ to decrypt every encrypted data key in parallel
+4. Ensure it is possible to efficiently limit decryption to specific CMKs
+   by passing the specific CMKs I want to use as part of my decryption call
+
+## Proposed Design
+
+The high level approach is to separate the existing AWS KMS keyring functionality into separate keyrings,
+with each keyring maintaining a relatively small surface area.
+To handle multiple strings identifying AWS KMS CMKs
+([key names](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#key-names))
+or multiple AWS regions,
+the proposal uses _keyring-producing operations_
+that produce a multi-keyring of smaller-scoped child AWS KMS keyrings.
+Each child AWS KMS keyring is tied to a single AWS SDK KMS service client (AWS SDK client)
+that is initialized before child keyring initialization.
+We refer to the keyrings that keyring-producing operations produce as _derived keyrings_.
+
+By simplifying the AWS KMS keyring,
+the basic building blocks for long-term maintainability and extensibility are established.
+ESDK developers can support additional customer use cases by writing new keyring-producing operations
+without needing to refactor existing code or update customer-facing APIs.
+Customers are given the same building blocks,
+which allows for additional customization and reduces the need for forked code.
+The smaller components reinforce an "easy to use, hard to misuse" mindset.
+
+### AWS KMS symmetric keyring
+
+For the purpose of this document,
+the proposed design’s _AWS KMS symmetric keyring_ will be referred to as the _AWS KMS keyring_
+for brevity.
+
+**Initialization**
+
+On keyring initialization, an AWS KMS symmetric keyring MUST define the following:
+
+- [Key Name](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#key-names)
+- AWS SDK KMS service client (AWS SDK client)
+
+On keyring initialization, an AWS KMS symmetric keyring MAY define the following:
+
+- Grant Tokens
+
+**Behavior**
+
+The AWS KMS symmetric keyring MUST...
+
+1. Be configured with a single key name
+   and a single AWS SDK client for that key name.
+   1. The AWS SDK client MUST be configured with credentials for an AWS principal
+      that passes an [AWS KMS Decrypt API](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html)
+      authorization check for the encrypted data key.
+2. Use the AWS SDK client provided at initialization for all AWS KMS API calls.
+3. Have an AWS KMS symmetric decryption contract
+   that ensures the encrypted data key’s [provider info](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/structures.md#key-provider-information)
```

Comment:
This means that on decrypt the provided EDK's provider info MUST match
the configured value.
As opposed to on encrypt the created EDK's provider info should match the configured value

Reply:
Yup! I'll add clarification. Will likely copy/paste the existing spec language from https://github.com/awslabs/aws-encryption-sdk-specification/blob/be6870b2513f4ca44bea8f2e0b5eb4808dba4365/framework/kms-keyring.md#key-names